pageserver/src/tenant/timeline.rs

```diff
@@ -2838,7 +2838,7 @@ impl Timeline {
             self.metrics
                 .resident_physical_size_gauge
                 .add(metadata.len());
-            updates.insert_historic(Arc::new(l));
+            assert!(updates.insert_historic(Arc::new(l)));
```

Comment:
Note we do have other callsites for the method.

I dislike assert usage here, with poisoned layermap I doubt we will be able to detach or ignore the tenant. I hope to look at it some day as well, we should `ignore` a tenant on poisoning.

`assert!` should definitely be used from the layermap unit tests I've added, or when loading the timeline.. Unsure about the rest.

Perhaps we could go with the assumption that we really shouldn't be generating duplicates.

Reply:
I am not sure that I addressed you concern (about poisoned layer map), but there is no more assert in this point.
Assert is moved down. But I still think that detection of duplicates means presence of some bug. So restart of pageserver after fixing this bug is the only acceptable behavior. In this case "poisoned tenant" is not a problem.